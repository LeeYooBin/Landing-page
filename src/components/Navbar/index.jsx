/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { v4 as uuidv4 } from 'uuid';
import Link from '../Link';

const Content = styled.div``;

const StyledBurger = styled.button`
  width: 2rem;
  height: 2rem;
  background: transparent;
  border: none;
  display: flex;
  justify-content: space-around;
  flex-flow: column nowrap;
  position: fixed;
  top: 2%;
  right: 3%;
  z-index: 20;

  div {
    width: 2rem;
    height: 0.25rem;
    border-radius: 10px;
    transform-origin: 1px;
    transition: all 0.3s linear;
    background-color: ${({ theme }) => theme.colors.gray4};

    &:nth-child(1) {
      transform: ${({ isOpen }) => (isOpen ? 'rotate(49deg)' : 'rotate(0)')};
    }

    &:nth-child(2) {
      transform: ${({ isOpen }) =>
        isOpen ? 'translateX(100%)' : 'translateX(0)'};
      opacity: ${({ isOpen }) => (isOpen ? 0 : 1)};
    }

    &:nth-child(3) {
      transform: ${({ isOpen }) => (isOpen ? 'rotate(-49deg)' : 'rotate(0)')};
    }
  }

  @media (min-width: 920px) {
    display: none;
  }
`;

const Menu = styled.nav`
  display: ${({ isOpen }) => (isOpen ? 'flex' : 'none')};
  width: auto;
  height: auto;
  min-height: 25vh;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 10%;
  background-color: ${({ theme }) => theme.colors.backgroundMenu};
  position: fixed;
  top: 0;
  right: 0;
  text-align: center;
  padding-top: 5rem;
  z-index: 15;

  @media (min-width: 920px) {
    position: static;
    display: flex;
    padding: 0;
    margin: 0;
    flex-direction: row;
    background: transparent;
    width: auto;
    justify-content: flex-start;
    min-height: auto;
  }
`;

const Navbar = ({ links = [] }) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Content>
      <StyledBurger isOpen={isOpen} onClick={() => setIsOpen(!isOpen)}>
        <div />
        <div />
        <div />
      </StyledBurger>
      <Menu isOpen={isOpen}>
        {links.map((item) => (
          <Link key={uuidv4()} link={item} onClick={() => setIsOpen(!isOpen)} />
        ))}
      </Menu>
    </Content>
  );
};

Navbar.propTypes = {
  links: PropTypes.arrayOf(PropTypes.string),
};

export default Navbar;
