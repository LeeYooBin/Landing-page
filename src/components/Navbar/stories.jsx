import Navbar from '.';

export default {
  title: 'Navbar',
  component: Navbar,
};

const links = ['about', 'missions', 'contact'];

export const Template = (args) => <Navbar {...args} links={links} />;
