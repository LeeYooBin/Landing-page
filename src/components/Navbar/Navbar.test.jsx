import { describe, it, expect } from '@jest/globals';
import { screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Navbar from '.';
import renderTheme from '../../styles/render-theme';

describe('Text component', () => {
  it('Should render with default values', () => {
    renderTheme(<Navbar />);
  });

  it('should render with the props without errors', () => {
    renderTheme(<Navbar links={['link1', 'link2', 'link3']} />);
    expect(screen.getByRole('navigation')).toBeInTheDocument();
  });

  it('should display the menu correctly when it is open', () => {
    renderTheme(<Navbar links={['link1', 'link2', 'link3']} />);
    expect(screen.getAllByRole('link')).toHaveLength(3);
  });

  it('verifies links rendering', () => {
    renderTheme(<Navbar links={['link1', 'link2', 'link3']} />);
    const linkElements = screen.getAllByRole('link');

    expect(linkElements).toHaveLength(3);
    expect(linkElements[0]).toHaveAttribute('href', '#link1');
    expect(linkElements[1]).toHaveAttribute('href', '#link2');
    expect(linkElements[2]).toHaveAttribute('href', '#link3');
  });

  it('should have menu closed by default only on mobile screens', () => {
    renderTheme(<Navbar links={['link1', 'link2', 'link3']} />);
    const isMobile = window.innerWidth < 920;
    const element = screen.getByRole('navigation');

    if (isMobile) {
      expect(element).not.toBeVisible();
    } else {
      expect(element).toBeVisible();
    }
  });

  it('should toggle menu state when burger/x icon is clicked and render the links correctly', () => {
    renderTheme(<Navbar links={['link1', 'link2', 'link3']} />);
    const isMobile = window.innerWidth < 920;
    const button = screen.getByRole('button', { name: '' });
    const menu = screen.getByRole('navigation');

    if (isMobile) {
      expect(menu).not.toBeVisible();

      fireEvent.click(button);
      const linkElements = screen.getAllByRole('link');
      expect(menu).toBeVisible();
      expect(linkElements).toHaveLength(3);
      expect(linkElements[0]).toHaveAttribute('href', '#link1');
      expect(linkElements[1]).toHaveAttribute('href', '#link2');
      expect(linkElements[2]).toHaveAttribute('href', '#link3');

      fireEvent.click(button);
      expect(menu).not.toBeVisible();
    }
  });

  it('tests the functionality of the links', () => {
    renderTheme(<Navbar links={['section1', 'section2', 'section3']} />);
    const links = screen.getAllByRole('link');

    links.forEach((link) => {
      expect(() => {
        fireEvent.click(link);
      }).not.toThrow();
    });
  });

  it('Should match snapshot', () => {
    const { container } = renderTheme(<Navbar />);
    expect(container).toMatchSnapshot();
  });
});
