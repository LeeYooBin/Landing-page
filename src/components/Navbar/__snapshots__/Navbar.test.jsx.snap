// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component Should match snapshot 1`] = `
.c0 {
  width: 2rem;
  height: 2rem;
  background: transparent;
  border: none;
  display: flex;
  justify-content: space-around;
  flex-flow: column nowrap;
  position: fixed;
  top: 2%;
  right: 3%;
  z-index: 20;
}

.c0 div {
  width: 2rem;
  height: 0.25rem;
  border-radius: 10px;
  transform-origin: 1px;
  transition: all 0.3s linear;
  background-color: #898A93;
}

.c0 div:nth-child(1) {
  transform: rotate(0);
}

.c0 div:nth-child(2) {
  transform: translateX(0);
  opacity: 1;
}

.c0 div:nth-child(3) {
  transform: rotate(0);
}

.c1 {
  display: none;
  width: auto;
  height: auto;
  min-height: 25vh;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 10%;
  background-color: #2d3142;
  position: fixed;
  top: 0;
  right: 0;
  text-align: center;
  padding-top: 5rem;
  z-index: 15;
}

@media (min-width: 920px) {
  .c0 {
    display: none;
  }
}

@media (min-width: 920px) {
  .c1 {
    position: static;
    display: flex;
    padding: 0;
    margin: 0;
    flex-direction: row;
    background: transparent;
    width: auto;
    justify-content: flex-start;
    min-height: auto;
  }
}

<div>
  <div
    class=""
  >
    <button
      class="c0"
    >
      <div />
      <div />
      <div />
    </button>
    <nav
      class="c1"
    />
  </div>
</div>
`;
